css/selectors/focus-visible-017.html

```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<meta charset="utf-8" />
+<title>CSS Test (Selectors): By default a programatic focus that matches :focus-visible shows an auto focus ring</title>
+<link rel="author" title="Manuel Rego Casasnovas" href="mailto:rego@igalia.com">
+<link rel="help" href="https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo" />
+<link rel="help" href="https://html.spec.whatwg.org/#phrasing-content-3" />
+<meta name="assert" content="This test checks that by default, if using JavaScript to focus an element triggers ':focus-visible' matching, then the element should show a focus ring with 'outline-style: auto'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<style>
+  @supports not (selector(:focus-visible)) {
+    :focus {
+      background-color: red;
+    }
+  }
+
+  :focus-visible {
+    background-color: lime;
+  }
+
+  :focus:not(:focus-visible) {
+    background-color: red;
+  }
+</style>
+
+<div id="target" tabindex="0">Target</div>
+<script>
+  async_test(function(t) {
+    target.addEventListener("focus", t.step_func(function() {
+      assert_equals(getComputedStyle(target).outlineStyle, "auto", `outline-style for ${target.tagName}#${target.id} should be auto`);
+      assert_equals(getComputedStyle(target).backgroundColor, "rgb(0, 255, 0)", `backgroundColor for ${target.tagName}#${target.id} should be lime`);
+      t.done();
+    }));
+    target.focus();
```

Comment:
To clarify my earlier comment: this line only triggers `:focus-visible` matching coincidentally, because we match `:focus-visible` by default when the page loads. Depending on that kind of unspecified behaviour makes me anxious. 

I explained in my earlier comment how to ensure that `.focus()` causes the newly focused element to match `:focus-visible`.

Reply:
I didn't realize this was not specified behavior... isn't that what https://wpt.fyi/results/css/selectors/focus-visible-010.html is testing? Or am I missing something else?

I understand now why you asked for adding more elements.

Reply:
Oh haha, I guess I forgot about that! Thanks for reminding me; in that case this is fine.

Reply:
(Might be worth adding a comment explaining that, in case I forget again.)

Reply:
TBH I cannot find this behavior defined on [the spec](https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo), so it's `focus-visible-010.html` that is wrong, or is the spec not updated?

The spec says:
* If the active element matches :focus-visible, and a script causes focus to move elsewhere, the newly focused element should match :focus-visible.
* Conversely, if the active element does not match :focus-visible, and a script causes focus to move elsewhere, the newly focused element should not match :focus-visible.

But doesn't mention the initial focus, when there's no active element before the focus. WDYT?

Reply:
I suspect we snuck it into tests without adding it to the spec. Do you think we should add it to the list of heuristics as well?

Reply:
I think it'd be nice adding it yeah, I can fill a CSSWG issue to do that. As Chromium and Firefox are already passing `focus-visible-010.html` so it doesn't need implementation changes.

Reply:
Thank you!

Reply:
Done: https://github.com/w3c/csswg-drafts/issues/5885

Thank you for the thorough reviews and all the feedback on these tests!